src/index-format/union_find.ml

```diff
@@ -0,0 +1,38 @@
+type 'a content =
+  | Root of { mutable value : 'a; mutable rank : int }
+  | Link of { mutable parent : 'a element }
+and 'a element = { mutable content : 'a content }
+
+let make value = { content = Root { value; rank = 0 } }
+
+let rec find x =
+  match x.content with
+  | Root _ -> x
+  | Link ({ parent; _ } as link) ->
+    let root = find parent in
+    if root <> parent then link.parent <- root;
```

Comment:
Maybe `!=` instead of `<>`, just to be sure? (I guess it's accidentally not a performance issue, because structural equality immediately sees either a physical equality or a difference of `Root`/`Link` constructor)

Reply:
Thanks, I chanded it.